File: 質數(試除法).c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#define MAX_NUM 1000000

int *primes, size;

bool isPrime(int n) {
    int limit = sqrt(n);
    for(int i = 0; i < size && primes[i] <= limit; ++i) {
        if(n % primes[i] == 0)
            return false;
    }
    return true;
}

void primeTrailDivision() {
    primes = (int*)malloc(sizeof(int));
    primes[0] = 2;
    size = 1;
    for(int n = 3; n <= MAX_NUM; n += 2) {
        if(isPrime(n)) {
            primes = (int*)realloc(primes, (size + 1) * sizeof(int));
            primes[size++] = n;
        }
    }
}

int main() {
    primeTrailDivision();
    int n;
    while(scanf("%d", &n) && n != 0)
        printf("%d is %sprime\n", n, isPrime(n) ? "" : "not ");
    free(primes);
    return 0;
}

File: 質數(試除法).java
import java.io.*;
import java.util.*;

public class Main {
    static int MAX_NUM = 1000000, primeCount;
    static boolean[] isComposite = new boolean[MAX_NUM + 1];
	static ArrayList<Integer> primes = new ArrayList<>();
	public static void main(String[] args) throws IOException {
	    primeTrailDivision();
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        StringBuilder output = new StringBuilder();
        int n;
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0)
            output.append(n).append(" is ").append(isPrime(n) ? "prime\n" : "not prime\n");
        System.out.print(output);
	}
	
	static boolean isPrime(int n) {
        int limit = (int)Math.sqrt(n);
        for(int i = 0; i < primeCount && primes.get(i) <= limit; ++i) {
            if(n % primes.get(i) == 0)
                return false;
        }
        return true;
    }
    
    static void primeTrailDivision() {
        primes.add(2);
        primeCount = 1;
        for(int n = 3; n <= MAX_NUM; n += 2) {
            if(isPrime(n)) {
                primes.add(n);
                ++primeCount;
            }
        }
    }
}
